components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesBookmarksStorage.kt

```diff
@@ -0,0 +1,177 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.storage.sync
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.withContext
+import mozilla.appservices.places.BookmarkFolder
+import mozilla.appservices.places.BookmarkItem
+import mozilla.appservices.places.BookmarkRoot
+import mozilla.appservices.places.BookmarkSeparator
+import mozilla.appservices.places.BookmarkTreeNode
+import mozilla.appservices.places.BookmarkUpdateInfo
+import mozilla.appservices.places.PlacesException
+import mozilla.appservices.places.PlacesReaderConnection
+import mozilla.appservices.places.PlacesWriterConnection
+import mozilla.components.concept.storage.BookmarksStorage
+import mozilla.components.concept.sync.AuthInfo
+import mozilla.components.concept.sync.SyncStatus
+import mozilla.components.concept.sync.SyncableStore
+
+/**
+ * Implementation of the [BookmarksStorage] which is backed by a Rust Places lib via [PlacesConnection].
+ */
+open class PlacesBookmarksStorage(context: Context) : BookmarksStorage, SyncableStore {
+
+    private val scope by lazy { CoroutineScope(IO) }
+    private val storageDir by lazy { context.filesDir }
+    @VisibleForTesting
+    internal open val places: Connection by lazy {
```

Comment:
While this will work correctly because `RustPlacesConnection` is a singleton, and because it caches its readers and writers, nothing in the current composition of the two storage layers - history and bookmarks - makes it clear that it's the same db underneath, and the same connection objects should be used. And the fact that you're passing in `storageDir` twice makes this potentially even more confusing.

I would factor out bits that call `RustPlacesConnection.init(...)` into an abstract base class, and extend it in `PlacesBookmarksStorage` and `PlacesHistoryStorage`. `PlacesStorage` seems like a good name for that base class.

Reply:
Good catch! This behavior belongs in a base class.